File: common/src/main/java/juuxel/adorn/lib/registry/Registered.java
package juuxel.adorn.lib.registry;

import net.minecraft.resources.ResourceKey;
import net.minecraft.core.Holder;

import java.util.function.Supplier;

@FunctionalInterface
public interface Registered<T> extends Supplier<T> {
    interface WithKey<R, T extends R> extends Registered<T> {
        ResourceKey<R> key();
        Holder<R> entry();
    }
}

File: common/src/main/kotlin/juuxel/adorn/lib/registry/Registered.kt
package juuxel.adorn.lib.registry

import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty

fun interface Registered<out T> : ReadOnlyProperty<Any?, T> {
    fun get(): T

    override fun getValue(thisRef: Any?, property: KProperty<*>): T = get()
}
